File: marketplace-backend/src/main/java/com/whatshop/marketplace/auth/service/AuthService.java
package com.whatshop.marketplace.auth.service;

import com.whatshop.marketplace.auth.dto.AuthResponse;
import com.whatshop.marketplace.auth.dto.LoginRequest;
import com.whatshop.marketplace.auth.dto.RegisterRequest;
import com.whatshop.marketplace.auth.entity.User;
import com.whatshop.marketplace.shared.response.ApiResponse;

public interface AuthService {
    ApiResponse<AuthResponse> register(RegisterRequest request);
    ApiResponse<AuthResponse> login(LoginRequest request);
    ApiResponse<AuthResponse> me(User user);
    ApiResponse<AuthResponse> refresh(User user);
}

File: marketplace-frontend/src/app/core/services/auth.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';

import { environment } from '../../../environments/environment';
import { ApiResponse } from '../models/api-response';
import { AuthResponse, LoginRequest, RegisterRequest } from '../models/auth';
import { TokenStorageService } from './token-storage.service';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly http = inject(HttpClient);
  private readonly tokenStorage = inject(TokenStorageService);
  private readonly authUrl = `${environment.apiUrl}/auth`;

  login(credentials: LoginRequest): Observable<ApiResponse<AuthResponse>> {
    return this.http.post<ApiResponse<AuthResponse>>(`${this.authUrl}/login`, credentials).pipe(
      tap(response => {
        if (response.success && response.data?.token) {
          this.tokenStorage.saveToken(response.data.token);
        }
      })
    );
  }

  register(payload: RegisterRequest): Observable<ApiResponse<AuthResponse>> {
    return this.http.post<ApiResponse<AuthResponse>>(`${this.authUrl}/register`, payload).pipe(
      tap(response => {
        if (response.success && response.data?.token) {
          this.tokenStorage.saveToken(response.data.token);
        }
      })
    );
  }
}